rust/scx_utils/src/load_balance.rs

```diff
@@ -0,0 +1,318 @@
+// Copyright (c) Meta Platforms, Inc. and affiliates.
+
+// This software may be used and distributed according to the terms of the
+// GNU General Public License version 2.
+
+//! # SCX Load Balancer
+//!
+//! A crate providing abstractions for implementing load balancing logic
+//! between scheduling domains.
+//!
+//! Initializing LoadBalancer
+//! -------------------------
+//!
+//! LoadBalancer objects are created from LoadAggregator objects. As the name
+//! suggests, LoadAggregator objects aggregate load passed by the caller, and
+//! create a LoadBalancer object that contains load and weights for each
+//! domain; adjusted for any infeasible weights.
+//!
+//! Note that the abstractions here are meant to be generic across schedulers.
+//! LoadAggregator assumes nothing about the scheduling domains being passed to
+//! it. For example, they may span layers defined in a scheduler, domains
+//! specified by the user via cpumask strings, or domains that span L3 caches.
+//! The crate simply aggregates load according to domain with IDs specified by
+//! the caller, computes global load sums, and adjusts any infeasible weights
+//! such that all infeasible scheduling entities still get their full duty
+//! cycle.
+//!
+//! Assume we're on a 16-core (32-CPU) host with two core complexes:
+//!
+//!```
+//!     let mut aggregator = LoadBalancer::aggregator(32, false);
+//!     aggregator.record_dom_load(0, 1, 1.0);
+//!     aggregator.record_dom_load(1, 1, 1.0);
+//!     // ...
+//!     aggregator.record_dom_load(31, 1, 1.0);
+//!     // ...
+//!     aggregator.record_dom_load(63, 1, 1.0);
+//!     aggregator.record_dom_load(64, 10000, 1.0);
+//!
+//!     let balancer = aggregator.calculate();
+//!
+//!     // Outputs: 66.06451612903226
+//!     info!("{}", balancer.global_load_sum());
+//!```
+//!
+//! For more details on the infeasible weights problem, please see the following
+//! document:
+//!
+//! https://drive.google.com/file/d/1fAoWUlmW-HTp6akuATVpMxpUpvWcGSAv
+
+use anyhow::bail;
+use anyhow::Result;
+use std::collections::BTreeMap;
+
+const MIN_WEIGHT: usize = 1;
+const MAX_WEIGHT: usize = 10000;
```

Comment:
Do we need to limit the max weight?

Reply:
Hmm, I suppose not. I guess there's no reason for us to stop callers from implementing their own weighting scheme beyond just e.g. `cpu.weight`. Will adjust.